File: app/src/main/java/com/example/bookinder/Register/RegisterActivity.java
package com.example.bookinder.Register;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.example.bookinder.Login.LoginActivity;
import com.example.bookinder.MainActivity;
import com.example.bookinder.ProfileExpansion.ProfileExpansion;
import com.example.bookinder.R;

import androidx.appcompat.app.AppCompatActivity;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class RegisterActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        Button Register = (Button) findViewById(R.id.signup_btn);
        Register.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                register(view);
            }});
    }

    public void register(View v) {
        EditText usernameView = findViewById(R.id.signup_email);
        EditText nameView = findViewById(R.id.signup_user_name);
        EditText phoneView = findViewById(R.id.signup_phone);
        EditText passwordView = findViewById(R.id.signup_password);
        EditText ageView = findViewById(R.id.signup_age);
        EditText confirmPasswordView = findViewById(R.id.signup_confirm_password);

        String username = usernameView.getText().toString().trim();
        String name = nameView.getText().toString().trim();
        String phone = phoneView.getText().toString().trim();
        String password = passwordView.getText().toString().trim();
        String confirmPassword = confirmPasswordView.getText().toString().trim();
        String age = ageView.getText().toString().trim();

        boolean flag = true;
        if(name.length() == 0) {
            nameView.setError("Name is required");
            flag = false;
        }
        if(username.length() == 0) {
            usernameView.setError("Email is required");
            flag = false;
        }
        if(password.length() == 0) {
            passwordView.setError("Password is required");
            flag = false;
        }
        if(confirmPassword.length() == 0) {
            confirmPasswordView.setError("You must confirm your password");
            flag = false;
        }
        if(!password.equals(confirmPassword)) {
            confirmPasswordView.setError("The password must be the same");
            flag = false;
        }
        if(age.length() == 0) {
            ageView.setError("Age is required");
            flag = false;
        }
        if(phone.length() == 0) {
            phoneView.setError("Phone number is required");
            flag = false;
        }


        if (!flag) {
            Toast.makeText(getApplicationContext(), "Something is wrong. Please check your inputs.", Toast.LENGTH_LONG).show();
        } else {
            JSONObject registrationForm = new JSONObject();
            try {
                registrationForm.put("subject", "register");
                registrationForm.put("name", name);
                registrationForm.put("email", username);
                registrationForm.put("phone", phone);
                registrationForm.put("password", password);
                registrationForm.put("age", age);
            } catch (JSONException e) {
                e.printStackTrace();
            }

            RequestBody body = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), registrationForm.toString());

            postRequest("http://192.168.1.169:5000", body);
        }
    }

    public void postRequest(String postUrl, RequestBody postBody) {
        OkHttpClient client = new OkHttpClient();

        final Request request = new Request.Builder()
                .url(postUrl)
                .post(postBody)
                .header("Accept", "application/json")
                .header("Content-Type", "application/json")
                .build();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                call.cancel();
                Log.d("FAIL", e.getMessage());

                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        TextView responseText = findViewById(R.id.responseTextRegister);
                        responseText.setText("Failed to Connect to Server. Please Try Again.");
                    }
                });
            }

            @Override
            public void onResponse(Call call, final Response response) {
                final TextView responseTextRegister = findViewById(R.id.responseTextRegister);
                try {
                    final String responseString = response.body().string().trim();
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            if (responseString.equals("success")) {
                                responseTextRegister.setText("Registration completed successfully.");
                                Intent intent = new Intent(RegisterActivity.this, ProfileExpansion.class);
                                startActivity(intent);
                                finish();
                            } else if (responseString.equals("Another user used this email. Please chose another email.")) {
                                responseTextRegister.setText("Email already exists. Please chose another email.");
                            } else {
                                responseTextRegister.setText("Something went wrong. Please try again later.");
                            }
                        }
                    });
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
    }
}

File: app/src/main/java/com/example/bookinder/Register/RegisterActivity.kt
package com.example.bookinder.Register

import android.content.Intent
import android.os.Bundle
import android.telephony.PhoneNumberUtils
import android.text.TextUtils
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.bookinder.Login.LoginActivity
import com.example.bookinder.R
import com.example.bookinder.api.RetrofitClient
import com.example.bookinder.models.DefaultResponse
import kotlinx.android.synthetic.main.activity_register.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class RegisterActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        move_to_login.setOnClickListener {
            startActivity(Intent(this@RegisterActivity, LoginActivity::class.java))
        }

        signup_btn.setOnClickListener {
            val username = signup_user_name.text.toString().trim()
            val email = signup_email.text.toString().trim()
            val password = signup_password.text.toString().trim()
            val confirm_password = signup_confirm_password.text.toString().trim()
            val age = signup_age.text.toString().trim()
            val phone_number = signup_phone.text.toString().trim()

            if(username.isEmpty()) {
                signup_user_name.error = "Username is required"
                signup_user_name.requestFocus()
                return@setOnClickListener
            }

            if(email.isEmpty()) {
                signup_email.error = "Email is required"
                signup_email.requestFocus()
                return@setOnClickListener
            }
            if (!email.isEmailValid()) {
                signup_email.error = "Email is not valid"
                signup_email.requestFocus()
                return@setOnClickListener
            }
            if(password.isEmpty()) {
                signup_password.error = "Password is required"
                signup_password.requestFocus()
                return@setOnClickListener
            }
            if (!isValidPassword(password)) {
                signup_password.error = "The password must include at least one upper and lower " +
                        "case letter, one digit and at least 4 characters"
                signup_password.requestFocus()
                return@setOnClickListener
            }
            if(confirm_password.isEmpty()) {
                signup_confirm_password.error = "You must confirm your password"
                signup_confirm_password.requestFocus()
                return@setOnClickListener
            }
            if (password != confirm_password) {
                signup_password.error = "Password is not match"
                signup_password.requestFocus()
                return@setOnClickListener
            }
            try {
                val age_number = age.toInt()
                if (age_number < 0 || age_number > 120) {
                    signup_age.error="invalid age"
                    signup_age.requestFocus()
                    return@setOnClickListener
                }
            } catch (e: NumberFormatException){
                signup_age.error="age must be a number"
                signup_age.requestFocus()
                return@setOnClickListener
            }
            if(!PhoneNumberUtils.isGlobalPhoneNumber(phone_number)) {
                signup_phone.error = "invalid phone number"
                signup_phone.requestFocus()
                return@setOnClickListener
            }

            RetrofitClient.instance.createUser(username,email,password,age,phone_number).
            enqueue(object: Callback<DefaultResponse>{
                override fun onResponse(call: Call<DefaultResponse>, response: Response<DefaultResponse>) {
                   Toast.makeText(applicationContext, response.body()?.message,Toast.LENGTH_LONG).show()
                }

                override fun onFailure(call: Call<DefaultResponse>, t: Throwable) {
                    Toast.makeText(applicationContext, t.message, Toast.LENGTH_LONG).show()
                }

            })
        }
    }

    fun isValidPassword(password: String?) : Boolean {
        password?.let {
            val passwordPattern = "^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&+=])(?=\\S+$).{4,}$"
            val passwordMatcher = Regex(passwordPattern)

            return passwordMatcher.find(password) != null
        } ?: return false
    }

    fun String.isEmailValid(): Boolean {
        return !TextUtils.isEmpty(this) && android.util.Patterns.EMAIL_ADDRESS.matcher(this).matches()
    }
}